### PR TITLE
Teach flutter msbuild for Windows

### PR DESCRIPTION
**Pull request by stuartmorgan-g**
## Description

Eliminates the need for a build.bat in the Windows build workflow, adding preliminary support for building using `msbuild`. The handling of `vcvars64.bat` may be refined in the future, but this serves as a starting point.

## Related Issues

https://github.com/flutter/flutter/issues/30706

## Tests

I added the following tests:

Validation of generate properties file and vcvars64.bat detection in existing workflow test.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by stuartmorgan** [comment]
Still needs tests, but the code itself is ready for a look.

File: packages/flutter_tools/lib/src/windows/msbuild_utils.dart [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  if (whereResult.exitCode == 0) {
+    return whereResult.stdout.trim();
+  }
+
+  return null;
+}
+
+/// Writes a property sheet (.props) file to expose all of the key/value
+/// pairs in [variables] as enivornment variables.
+Future<void> writePropertySheet(
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/windows/msbuild_utils.dart]: We should add a test that the xml has the correct structure. This should be fairly easy to read from the MemoryFileSystem

**Review by jonahwilliams — APPROVED** [review]
LGTM with nit

File: packages/flutter_tools/lib/src/windows/msbuild_utils.dart [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    builder.element('ImportGroup', nest: () {
+      builder.attribute('Label', 'PropertySheets');
+    });
+    _addUserMacros(builder, variables);
+    builder.element('PropertyGroup');
+    builder.element('ItemDefinitionGroup');
+    _addItemGroup(builder, variables);
+  });
+
+  await propertySheetFile.create(recursive: true);
```
[the remark this answers is not in the document]
Comment (stuartmorgan) [reply packages/flutter_tools/lib/src/windows/msbuild_utils.dart]: Done.

File: packages/flutter_tools/lib/src/windows/msbuild_utils.dart [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+import '../base/process_manager.dart';
+
+/// Returns the path to an installed vcvars64.bat script if found, or null.
+Future<String> findVcvars() async {
+  final String programDir = platform.environment['PROGRAMFILES(X86)'];
+  final String pathPrefix = fs.path.join(programDir, 'Microsoft Visual Studio');
+  const String vcvarsScriptName = 'vcvars64.bat';
+  final String pathSuffix =
+      fs.path.join('VC', 'Auxiliary', 'Build', vcvarsScriptName);
+  final List<String> years = <String>['2017', '2019'];
```
[the remark this answers is not in the document]
Comment (stuartmorgan) [reply packages/flutter_tools/lib/src/windows/msbuild_utils.dart]: Done.

File: packages/flutter_tools/lib/src/windows/msbuild_utils.dart (hunk 1 above)
Comment (stuartmorgan) [reply packages/flutter_tools/lib/src/windows/msbuild_utils.dart]: Done. (We may want to add more over time, but there's some initial validation now.)

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Pull request merged**